### PR TITLE
Port verifer to tornado release 6.0.x

### PR DESCRIPTION
**Pull request by lukehinds**
`callback` has been depreciated from tornado 6.0

This code changes the `invoke_get_quote` from a callback to a
while loop

Resolves: #196

### CODE REVIEW

**Comment by lukehinds** [comment]
This actually works well, but I get a small hiccup where nonce is either a string or an int

ima

File: keylime/cloud_verifier_tornado.py [hunk 1]
```diff
@@ -293,77 +294,71 @@ def put(self):
...
-
-    def on_get_quote_response(self, agent, url, response):
-        if agent is None:
-            raise Exception("agent deleted while being processed")
-        if response.error:
-            # this is a connection error, retry get quote
-            if isinstance(response.error, IOError) or (isinstance(response.error, tornado.web.HTTPError) and response.error.code == 599):
-                self.process_agent(agent, cloud_verifier_common.CloudAgent_Operational_State.GET_QUOTE_RETRY)
+        while True:
+            params = f'/quotes/integrity?nonce={params["nonce"]}&mask={params["mask"]}&vmask={params["vmask"]}&partial={partial_req}'
```
Comment (jetwhiz) [inline keylime/cloud_verifier_tornado.py]: It looks like the existing `params` variable (a dictionary) is being overwritten here (with a string).  Recommend renaming to `path` (and updating below to `params=path`).

Comment (lukehinds) [reply keylime/cloud_verifier_tornado.py]: Ah yes, good catch!

Comment (jetwhiz) [reply keylime/cloud_verifier_tornado.py]: Does this solve the issue you described above, or is that unrelated?

**Comment by nabilschear** [comment]
if i'm reading this right, this means that the verifier will block waiting for a quote from the agent?  If so, this means that the whole verifier I/O loop will be blocked for quite some time waiting on a hardware TPM (500-1000ms) on the agent.  So, the verifier will be unresponsive during that time, to adding new agents or getting status of old ones, etc..  Since we've got multiple tornado I/O loop processes, it won't stop completely.  But as soon as we've got more agents than we have processes, then the verifier is going to lock up.  I think that we should look at the nested async approach more closely.  can we package it for fedora?  how hard would that be?

**Comment by lukehinds** [comment]
@nabilschear 

I can set it up as async, as it happens I already have the code here..

**Comment by lukehinds** [comment]
@nabilschear https://github.com/keylime/keylime/pull/200/commits/990b713cec9a53498e3f0290cd63b21c0545ebfa

**Comment by lukehinds** [comment]
actually, seems there is an issue here still:

```
2019-11-04 17:11:48.758 - keylime.tpm - INFO - Checking IMA measurement list...
2019-11-04 17:11:49.060 - keylime.cloudverifier - ERROR - Polling thread error: nothing should ever fall out of this!
2019-11-04 17:11:49.060 - keylime.cloudverifier - ERROR - nothing should ever fall out of this!
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/keylime-1.2-py3.7.egg/keylime/cloud_verifier_tornado.py", line 461, in process_agent
    raise Exception("nothing should ever fall out of this!")
Exception: nothing should ever fall out of this!
2019-11-04 17:11:49.202 - keylime.tpm - INFO - Checking IMA measurement list...
2019-11-04 17:11:49.464 - keylime.cloudverifier - ERROR - Polling thread error: nothing should ever fall out of this!
2019-11-04 17:11:49.465 - keylime.cloudverifier - ERROR - nothing should ever fall out of this!
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/keylime-1.2-py3.7.egg/keylime/cloud_verifier_tornado.py", line 461, in process_agent
    raise Exception("nothing should ever fall out of this!")
Exception: nothing should ever fall out of this!
```

File: keylime/cloud_verifier_tornado.py [hunk 2]
```diff
@@ -433,9 +438,9 @@ def process_agent(self, agent, new_operational_state):
                         cloud_verifier_common.notifyError(agent,'comm_error')
                     else:
                         logger.debug("Communication error for new agent.  no notification will be sent")
-                    self.process_agent(agent, cloud_verifier_common.CloudAgent_Operational_State.FAILED)
+                    await self.process_agent(agent, cloud_verifier_common.CloudAgent_Operational_State.FAILED)
                 else:
-                    cb = functools.partial(self.invoke_get_quote, agent, True)
+                    cb = functools.partial(await self.invoke_get_quote, agent, True)
```
Comment (jetwhiz) [inline keylime/cloud_verifier_tornado.py]: I think this might be the issue, since we're `await`ing a function that's being passed in (and not executed).

**Comment by lukehinds** [comment]
Just an update, it's close to working now, but somehow the operational state is wrong (yet to determine how).

```
2019-11-06 13:19:12.985 - keylime.registrar_client - WARNING - CAUTION: using default password for private key, please set private_key_pw to a strong password
2019-11-06 13:19:13.024 - keylime.tpm2 - INFO - TPM2-TOOLS Version: 3.2.0
2019-11-06 13:19:13.049 - keylime.tpm - INFO - Checking IMA measurement list...
main_agent_operational_state: 3
main_agent_operational_state: 5
2019-11-06 13:19:14.208 - keylime.tpm - INFO - Checking IMA measurement list...
main_agent_operational_state: 5
2019-11-06 13:19:14.638 - keylime.tpm - INFO - Checking IMA measurement list...
main_agent_operational_state: 5
2019-11-06 13:19:15.210 - keylime.tpm - INFO - Checking IMA measurement list...
main_agent_operational_state: 5
2019-11-06 13:19:15.692 - keylime.tpm - INFO - Checking IMA measurement list...
main_agent_operational_state: 5
2019-11-06 13:19:16.141 - keylime.tpm2 - ERROR - Error verifying quote: Command: tpm2_checkquote -c /tmp/tmpml2e80sq -m /tmp/tmp25elbxav -s /tmp/tmppslqlv5a -p /tmp/tmpxm68p797 -G sha256 -q 6a4769754d5a46464a6142384d69466f38335463 returned 1, expected 0, output [b'ERROR: Error validating nonce from quote\n', b'ERROR: Verify signature failed!\n', b'ERROR: Unable to run tpm2_checkquote\n', b'pcrs:\n', b'  sha256:\n', b'    15: 0x0000000000000000000000000000000000000000000000000000000000000000\n', b'    22: 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\n', b'  sha1:\n', b'    10: 0xA58ADD6790FC45E4B3393452FC84FCA4B44E262E\n', b'calcDigest: 4afd23281251611f8129a9f775f604ac298cc26ea227547ec61e6311cbe45305\n', b'msgDigest: 3f2866845cb4c94ae4686d49e1297046fe0a71f098671b8ccea1470f7e0038a2\n', b'sigBuffer: 39500ec2959c50b3783f0d794701eecf719d889e20010e1f5ac6e22e2262dffc00e8fb88ddaa487a3312e58bdc33cd429f9cd6a15c3112cde17374bd5dabc62d098d3e7f3d436702a16831b37245e8067aee63eab264a16cf4f85835885d433aa472b0d8ad05992897625912ee0a240d5c9078f2d4cdcd8ccd62dc3f115bfca2367d6eb3082dbe3464a733d88da3ac8f4d296d6cd43c912ed65971d2c4793bc62b360a5cc17ea9457fbc6860e9d0d895b3d12a1d7d05e9b2a5a72cb5d7f5eb343691394879347a562b837e339fcd4e4edcfc14491fa28e42c02d1a93a42d7f70b80efb75960c987451ef4055ee141f6fa0763a5e8a61427e79d8679095cb7220\n']
2019-11-06 13:19:16.142 - keylime.tpm2 - ERROR - Command: tpm2_checkquote -c /tmp/tmpml2e80sq -m /tmp/tmp25elbxav -s /tmp/tmppslqlv5a -p /tmp/tmpxm68p797 -G sha256 -q 6a4769754d5a46464a6142384d69466f38335463 returned 1, expected 0, output [b'ERROR: Error validating nonce from quote\n', b'ERROR: Verify signature failed!\n', b'ERROR: Unable to run tpm2_checkquote\n', b'pcrs:\n', b'  sha256:\n', b'    15: 0x0000000000000000000000000000000000000000000000000000000000000000\n', b'    22: 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\n', b'  sha1:\n', b'    10: 0xA58ADD6790FC45E4B3393452FC84FCA4B44E262E\n', b'calcDigest: 4afd23281251611f8129a9f775f604ac298cc26ea227547ec61e6311cbe45305\n', b'msgDigest: 3f2866845cb4c94ae4686d49e1297046fe0a71f098671b8ccea1470f7e0038a2\n', b'sigBuffer: 39500ec2959c50b3783f0d794701eecf719d889e20010e1f5ac6e22e2262dffc00e8fb88ddaa487a3312e58bdc33cd429f9cd6a15c3112cde17374bd5dabc62d098d3e7f3d436702a16831b37245e8067aee63eab264a16cf4f85835885d433aa472b0d8ad05992897625912ee0a240d5c9078f2d4cdcd8ccd62dc3f115bfca2367d6eb3082dbe3464a733d88da3ac8f4d296d6cd43c912ed65971d2c4793bc62b360a5cc17ea9457fbc6860e9d0d895b3d12a1d7d05e9b2a5a72cb5d7f5eb343691394879347a562b837e339fcd4e4edcfc14491fa28e42c02d1a93a42d7f70b80efb75960c987451ef4055ee141f6fa0763a5e8a61427e79d8679095cb7220\n']
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/keylime-1.2-py3.7.egg/keylime/tpm2.py", line 1052, in check_quote
    retDict = self.__check_quote_c(aikFile.name, nonce, quoteFile.name, sigFile.name, pcrFile.name, hash_alg)
  File "/usr/local/lib/python3.7/site-packages/keylime-1.2-py3.7.egg/keylime/tpm2.py", line 999, in __check_quote_c
    retDict = self.__run(command.format(**cmdargs), lock=False)
  File "/usr/local/lib/python3.7/site-packages/keylime-1.2-py3.7.egg/keylime/tpm2.py", line 290, in __run
    raise Exception("Command: %s returned %d, expected %d, output %s"%(cmd, code, expectedcode, retout))
Exception: Command: tpm2_checkquote -c /tmp/tmpml2e80sq -m /tmp/tmp25elbxav -s /tmp/tmppslqlv5a -p /tmp/tmpxm68p797 -G sha256 -q 6a4769754d5a46464a6142384d69466f38335463 returned 1, expected 0, output [b'ERROR: Error validating nonce from quote\n', b'ERROR: Verify signature failed!\n', b'ERROR: Unable to run tpm2_checkquote\n', b'pcrs:\n', b'  sha256:\n', b'    15: 0x0000000000000000000000000000000000000000000000000000000000000000\n', b'    22: 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\n', b'  sha1:\n', b'    10: 0xA58ADD6790FC45E4B3393452FC84FCA4B44E262E\n', b'calcDigest: 4afd23281251611f8129a9f775f604ac298cc26ea227547ec61e6311cbe45305\n', b'msgDigest: 3f2866845cb4c94ae4686d49e1297046fe0a71f098671b8ccea1470f7e0038a2\n', b'sigBuffer: 39500ec2959c50b3783f0d794701eecf719d889e20010e1f5ac6e22e2262dffc00e8fb88ddaa487a3312e58bdc33cd429f9cd6a15c3112cde17374bd5dabc62d098d3e7f3d436702a16831b37245e8067aee63eab264a16cf4f85835885d433aa472b0d8ad05992897625912ee0a240d5c9078f2d4cdcd8ccd62dc3f115bfca2367d6eb3082dbe3464a733d88da3ac8f4d296d6cd43c912ed65971d2c4793bc62b360a5cc17ea9457fbc6860e9d0d895b3d12a1d7d05e9b2a5a72cb5d7f5eb343691394879347a562b837e339fcd4e4edcfc14491fa28e42c02d1a93a42d7f70b80efb75960c987451ef4055ee141f6fa0763a5e8a61427e79d8679095cb7220\n']
main_agent_operational_state: 3
2019-11-06 13:19:16.424 - keylime.cloudverifier - WARNING - agent D432FBB3-D2F1-4A97-9EF7-75BD81C00000 failed, stopping polling
2019-11-06 13:19:16.446 - keylime.tpm - INFO - Checking IMA measurement list...
main_agent_operational_state: 9
2019-11-06 13:19:16.936 - keylime.tpm2 - ERROR - Error verifying quote: Command: tpm2_checkquote -c /tmp/tmpoykp4t63 -m /tmp/tmp4i9iz3av -s /tmp/tmppt4lu7u0 -p /tmp/tmpqzutgiv_ -G sha256 -q 6a4769754d5a46464a6142384d69466f38335463 returned 1, expected 0, output [b'ERROR: Error validating nonce from quote\n', b'ERROR: Verify signature failed!\n', b'ERROR: Unable to run tpm2_checkquote\n', b'pcrs:\n', b'  sha256:\n', b'    15: 0x0000000000000000000000000000000000000000000000000000000000000000\n', b'    22: 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\n', b'  sha1:\n', b'    10: 0xA58ADD6790FC45E4B3393452FC84FCA4B44E262E\n', b'calcDigest: 4afd23281251611f8129a9f775f604ac298cc26ea227547ec61e6311cbe45305\n', b'msgDigest: 1855362a6349fd53855cd73351942b2f66b1e3f5d16c6a7c2c7f2e2dadab880a\n', b'sigBuffer: 8ea8de0d2a6eae81518805f5d88b0dc7d59512a685eb5430fb7a6e2db14e20a7714fa2a82f1c4fde53910de0ebdec2eaca50910d708bc05758fd852d763c6b18057a2fcae92d7b7a27184c659847178cb1219a7acc5417420263355dccf7d920a7da27cb7f83c6a9f81c60271d8710e705ade49364aa58f8873a8ceacd2e4799e61b8d971c041077b760d860b6e489d50456f6619b1333f306c81fd17e073174750e4aa7f20259570c0a997c1cfd5421f63eab13df4807ca1fe6aa42bdda2f75115c7595f7e75cd55f91bd95cd14728362d5a645cc619b7128d89b202835754902aca6d207fe05aabb2e427d5e91cce7dddc640a373102c562c00bdd70c4ed65\n']
2019-11-06 13:19:16.936 - keylime.tpm2 - ERROR - Command: tpm2_checkquote -c /tmp/tmpoykp4t63 -m /tmp/tmp4i9iz3av -s /tmp/tmppt4lu7u0 -p /tmp/tmpqzutgiv_ -G sha256 -q 6a4769754d5a46464a6142384d69466f38335463 returned 1, expected 0, output [b'ERROR: Error validating nonce from quote\n', b'ERROR: Verify signature failed!\n', b'ERROR: Unable to run tpm2_checkquote\n', b'pcrs:\n', b'  sha256:\n', b'    15: 0x0000000000000000000000000000000000000000000000000000000000000000\n', b'    22: 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\n', b'  sha1:\n', b'    10: 0xA58ADD6790FC45E4B3393452FC84FCA4B44E262E\n', b'calcDigest: 4afd23281251611f8129a9f775f604ac298cc26ea227547ec61e6311cbe45305\n', b'msgDigest: 1855362a6349fd53855cd73351942b2f66b1e3f5d16c6a7c2c7f2e2dadab880a\n', b'sigBuffer: 8ea8de0d2a6eae81518805f5d88b0dc7d59512a685eb5430fb7a6e2db14e20a7714fa2a82f1c4fde53910de0ebdec2eaca50910d708bc05758fd852d763c6b18057a2fcae92d7b7a27184c659847178cb1219a7acc5417420263355dccf7d920a7da27cb7f83c6a9f81c60271d8710e705ade49364aa58f8873a8ceacd2e4799e61b8d971c041077b760d860b6e489d50456f6619b1333f306c81fd17e073174750e4aa7f20259570c0a997c1cfd5421f63eab13df4807ca1fe6aa42bdda2f75115c7595f7e75cd55f91bd95cd14728362d5a645cc619b7128d89b202835754902aca6d207fe05aabb2e427d5e91cce7dddc640a373102c562c00bdd70c4ed65\n']
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/keylime-1.2-py3.7.egg/keylime/tpm2.py", line 1052, in check_quote
    retDict = self.__check_quote_c(aikFile.name, nonce, quoteFile.name, sigFile.name, pcrFile.name, hash_alg)
  File "/usr/local/lib/python3.7/site-packages/keylime-1.2-py3.7.egg/keylime/tpm2.py", line 999, in __check_quote_c
    retDict = self.__run(command.format(**cmdargs), lock=False)
  File "/usr/local/lib/python3.7/site-packages/keylime-1.2-py3.7.egg/keylime/tpm2.py", line 290, in __run
    raise Exception("Command: %s returned %d, expected %d, output %s"%(cmd, code, expectedcode, retout))
Exception: Command: tpm2_checkquote -c /tmp/tmpoykp4t63 -m /tmp/tmp4i9iz3av -s /tmp/tmppt4lu7u0 -p /tmp/tmpqzutgiv_ -G sha256 -q 6a4769754d5a46464a6142384d69466f38335463 returned 1, expected 0, output [b'ERROR: Error validating nonce from quote\n', b'ERROR: Verify signature failed!\n', b'ERROR: Unable to run tpm2_checkquote\n', b'pcrs:\n', b'  sha256:\n', b'    15: 0x0000000000000000000000000000000000000000000000000000000000000000\n', b'    22: 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\n', b'  sha1:\n', b'    10: 0xA58ADD6790FC45E4B3393452FC84FCA4B44E262E\n', b'calcDigest: 4afd23281251611f8129a9f775f604ac298cc26ea227547ec61e6311cbe45305\n', b'msgDigest: 1855362a6349fd53855cd73351942b2f66b1e3f5d16c6a7c2c7f2e2dadab880a\n', b'sigBuffer: 8ea8de0d2a6eae81518805f5d88b0dc7d59512a685eb5430fb7a6e2db14e20a7714fa2a82f1c4fde53910de0ebdec2eaca50910d708bc05758fd852d763c6b18057a2fcae92d7b7a27184c659847178cb1219a7acc5417420263355dccf7d920a7da27cb7f83c6a9f81c60271d8710e705ade49364aa58f8873a8ceacd2e4799e61b8d971c041077b760d860b6e489d50456f6619b1333f306c81fd17e073174750e4aa7f20259570c0a997c1cfd5421f63eab13df4807ca1fe6aa42bdda2f75115c7595f7e75cd55f91bd95cd14728362d5a645cc619b7128d89b202835754902aca6d207fe05aabb2e427d5e91cce7dddc640a373102c562c00bdd70c4ed65\n']
main_agent_operational_state: 9
2019-11-06 13:19:17.210 - keylime.cloudverifier - WARNING - agent D432FBB3-D2F1-4A97-9EF7-75BD81C00000 failed, stopping polling
2019-11-06 13:19:17.237 - keylime.tpm - INFO - Checking IMA measurement list...
main_agent_operational_state: 9
```

**Comment by lukehinds** [comment]
@jetwhiz @nabilschear 

This is a full debug of the current situation, somehow the operational state between get quote and provide v falls out of sync and we end up with a loop. I put in some debug statements to track through the whole code flow

1. We start in `process_agent` and fall into the followng condition:

https://github.com/keylime/keylime/blob/6138c45a37823b6c321b47a22ca49ef736bfb6e1/keylime/cloud_verifier_tornado.py#L408-L413

2. We then move to `invoke_get_quote` and match the `else` clause and `if agent[`provide_v]`

https://github.com/keylime/keylime/blob/6138c45a37823b6c321b47a22ca49ef736bfb6e1/keylime/cloud_verifier_tornado.py#L322-L336

3. Next `process_agent` where we match operational states for `GET_QUOTE` and `PROVIDE_V`

https://github.com/keylime/keylime/blob/6138c45a37823b6c321b47a22ca49ef736bfb6e1/keylime/cloud_verifier_tornado.py#L415-L419

4. Next we move to `invoke_provide_v` where we fall into the else statement on line # 358

https://github.com/keylime/keylime/blob/6138c45a37823b6c321b47a22ca49ef736bfb6e1/keylime/cloud_verifier_tornado.py#L340-L359

5. We then move to `process_agent` again where we match `PROVIDE_V` and both `main_agent_operational_state` and `new_operational_state` match `GET_QUOTE`

https://github.com/keylime/keylime/blob/6138c45a37823b6c321b47a22ca49ef736bfb6e1/keylime/cloud_verifier_tornado.py#L421-L435

6. We now move back to `invoke_get_quote` where we again call `if agent['provide_V']` which is where we get into a loop

https://github.com/keylime/keylime/blob/6138c45a37823b6c321b47a22ca49ef736bfb6e1/keylime/cloud_verifier_tornado.py#L322-L336

7. We branch slightly differently to step 2 and now fall outside of the whole condition and end up at "nothing should ever fall out of this!" and the "polling thread error"

https://github.com/keylime/keylime/blob/6138c45a37823b6c321b47a22ca49ef736bfb6e1/keylime/cloud_verifier_tornado.py#L467-L471

```
2019-11-06 14:02:46.889 - keylime.cloudverifier - ERROR - Polling thread error: nothing should ever fall out of this!
2019-11-06 14:02:46.889 - keylime.cloudverifier - ERROR - nothing should ever fall out of this!
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/keylime-1.2-py3.7.egg/keylime/cloud_verifier_tornado.py", line 494, in process_agent
    raise Exception("nothing should ever fall out of this!")
Exception: nothing should ever fall out of this!
DEBUG 326
2019-11-06 14:02:48.124 - keylime.tpm2 - ERROR - Error verifying quote: Command: tpm2_checkquote -c /tmp/tmpakocfy0i -m /tmp/tmps2he9a1h -s /tmp/tmp9z0dwkcs -p /tmp/tmpekq43euz -G sha256 -q 73616f4e5a6a57425557427976447941325a414b returned 1, expected 0, output [b'ERROR: Error validating nonce from quote\n', b'ERROR: Verify signature failed!\n', b'ERROR: Unable to run tpm2_checkquote\n', b'pcrs:\n', b'  sha256:\n', b'    15: 0x0000000000000000000000000000000000000000000000000000000000000000\n', b'    22: 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\n', b'  sha1:\n', b'    10: 0xA58ADD6790FC45E4B3393452FC84FCA4B44E262E\n', b'calcDigest: 4afd23281251611f8129a9f775f604ac298cc26ea227547ec61e6311cbe45305\n', b'msgDigest: 3bc121ad74cf05ea881070da6f5646999c85e31fd011334369389f2018d6f49c\n', b'sigBuffer: 69ad57406ca3fe6719f830653596b59d0ec907cd15cb3a46342726bc0db66c08ae379523365242a610c64e913f0e78fa772659aba6364f5bff0ac2e84df7b6c82454e8ad158c6a9d32855ff564bac100ef3833bd1e1f9605690a80c99400a080f4bcef54f9168254de3956db52be1061c1ff956f896240ec1bba60fad773e5e72aac6f07185b37058ec5878f4a98799820f34d33e25a3514c6614402bfe90ffacba1518c126d45b4ca3517fe6c987d3f7845d0585e76e570173437fbb2efbadccfa8c063d660a65794c7cb5a58139e6880135c54a38b4e8c11fa458920c6d202dc2a500af1cbe3d894f05b1ff9a85c7b248113d992613b0a4dda7a45e576706a\n']
2019-11-06 14:02:48.125 - keylime.tpm2 - ERROR - Command: tpm2_checkquote -c /tmp/tmpakocfy0i -m /tmp/tmps2he9a1h -s /tmp/tmp9z0dwkcs -p /tmp/tmpekq43euz -G sha256 -q 73616f4e5a6a57425557427976447941325a414b returned 1, expected 0, output [b'ERROR: Error validating nonce from quote\n', b'ERROR: Verify signature failed!\n', b'ERROR: Unable to run tpm2_checkquote\n', b'pcrs:\n', b'  sha256:\n', b'    15: 0x0000000000000000000000000000000000000000000000000000000000000000\n', b'    22: 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\n', b'  sha1:\n', b'    10: 0xA58ADD6790FC45E4B3393452FC84FCA4B44E262E\n', b'calcDigest: 4afd23281251611f8129a9f775f604ac298cc26ea227547ec61e6311cbe45305\n', b'msgDigest: 3bc121ad74cf05ea881070da6f5646999c85e31fd011334369389f2018d6f49c\n', b'sigBuffer: 69ad57406ca3fe6719f830653596b59d0ec907cd15cb3a46342726bc0db66c08ae379523365242a610c64e913f0e78fa772659aba6364f5bff0ac2e84df7b6c82454e8ad158c6a9d32855ff564bac100ef3833bd1e1f9605690a80c99400a080f4bcef54f9168254de3956db52be1061c1ff956f896240ec1bba60fad773e5e72aac6f07185b37058ec5878f4a98799820f34d33e25a3514c6614402bfe90ffacba1518c126d45b4ca3517fe6c987d3f7845d0585e76e570173437fbb2efbadccfa8c063d660a65794c7cb5a58139e6880135c54a38b4e8c11fa458920c6d202dc2a500af1cbe3d894f05b1ff9a85c7b248113d992613b0a4dda7a45e576706a\n']
```

**Comment by nabilschear** [comment]
@lukehinds i did a little poking around, and I think we can solve this problem by using ensure_future and background tasks rather than await.  I'm not good enough at git to check this into your branch, so here's the tornado.py file in a gist https://gist.github.com/nabilschear/8e36f11bf28270a4cbe6ec8ba11112c4 This does appear to work, though there's one place where it may make sense to await or move that code into process_agent.  I think that the problems that @cjustacoder is having are related to all of this

**Comment by cjustacoder** [comment]
Hi, Nabil @nabilschear . This one seems like the agent code, not the verifier code. I'm a little confused about this.

**Comment by jetwhiz** [comment]
> Hi, Nabil @nabilschear . This one seems like the agent code, not the verifier code. I'm a little confused about this.

It looks like this has been fixed in the gist now

**Comment by cjustacoder** [comment]
> 
> 
> > Hi, Nabil @nabilschear . This one seems like the agent code, not the verifier code. I'm a little confused about this.
> 
> It looks like this has been fixed in the gist now

Yes, it is now, maybe I got some error on my side. Thanks for the update! @nabilschear @jetwhiz

**Comment by lukehinds** [comment]
@nabilschear , thank you so much. Just running it now it looks to be working fine to me. Well done on trying `ensure_future` , I was having a lot of trouble trying to work out how the various parts of asyncio could be used and would not clash to the tornados implementation. 

I will apply you gist to my PR and we can then move back to the Fedora packaging.

**Comment by nabilschear** [comment]
@lukehinds i need to ensure that revocations are still going out and a few other testing bits too.

**Comment by cjustacoder** [comment]
@lukehinds @nabilschear, I fall behind this process. Is there any update I need to catch up to get this work?

**Comment by lukehinds** [comment]
@nabilschear @jetwhiz  - ready for review now, once approved we can merge and I will jump back onto packaging.

**Comment by lukehinds** [comment]
> @lukehinds @nabilschear, I fall behind this process. Is there any update I need to catch up to get this work? What else should I update besides this tornado file?

After this has merged, you be able to merge the changes into your local branch. It should not be to complex, there are mainly some additional async calls to add.

**Comment by cjustacoder** [comment]
> 
> 
> > @lukehinds @nabilschear, I fall behind this process. Is there any update I need to catch up to get this work? What else should I update besides this tornado file?
> 
> After this has merged, you be able to merge the changes into your local branch. It should not be to complex, there are mainly some additional async calls to add.

Thanks, should I upgrade my python tornado package also? Has the callback functionality already invalid now?

**Comment by lukehinds** [comment]
@cjustacoder , you won't need to update, version 5.x is fine. This change is for stuff removed, rather than added.

**Comment by cjustacoder** [comment]
Why the state assignment 'agent['operational_state'] = cloud_verifier_common.CloudAgent_Operational_State.GET_QUOTE' is missing here: 
https://github.com/keylime/keylime/blob/09ec0145cc078ecd6f8a8f2eafa7537bdb6385e9/keylime/cloud_verifier_tornado.py#L297

**Comment by lukehinds** [comment]
> Why the state assignment `agent['operational_state'] = cloud_verifier_common.CloudAgent_Operational_State.GET_QUOTE` is missing here (in L298):
> 
> https://github.com/keylime/keylime/blob/09ec0145cc078ecd6f8a8f2eafa7537bdb6385e9/keylime/cloud_verifier_tornado.py#L294-L299

Well spotted. I have squashed commits now, so I can't see the history. @nabilschear I am not sure if this is missing from you copying my file into the gist (and I mistakenly removed it) or it was removed as part of your change to introduce ensure future calls?

**Comment by nabilschear** [comment]
I did remove that line from the `invoke_get_quote` function on purpose.   I had moved all the changes to that var into the `process_agent` function so that it would be clearer how the state machine operates.  All calls to `invoke_get_quote` do set the state to `GET_QUOTE` before calling the function.

**Comment by lukehinds** [comment]
> I did remove that line from the `invoke_get_quote` function on purpose. I had moved all the changes to that var into the `process_agent` function so that it would be clearer how the state machine operates. All calls to `invoke_get_quote` do set the state to `GET_QUOTE` before calling the function.

That's fine for me then.

**Comment by cjustacoder** [comment]
Thanks, it clear for me, too.

**Comment by lukehinds** [comment]
hi @nabilschear , did you perchance get to test the revocation checks you wanted to make?

I am quite keen to merge this one as soon as we can, as its a key need for landing packaging into Fedora (Fedora Rawhide has Tornado 6 packaged).

**Comment by nabilschear** [comment]
sorry, let me try that now

**Comment by nabilschear** [comment]
@lukehinds revocation is fine.  though i did run into some other bugs along the way.  0mq is a real pain unfortunately.

**Comment by nabilschear** [comment]
@lukehinds i think we also need to update codacy to use pylint for python3 instead of 2

**Comment by lukehinds** [comment]
Thanks for you work @nabilschear , squashed the commits and will give a quick test, we should then be able to merge and get back to Fedora packaging yay!

**Pull request merged**